### PR TITLE
feat: zkSync Extensions

### PR DESCRIPTION
**Pull request by mariopil**
### Focus of this PR
This PR focuses on adding zkSync Extensions and making various updates and additions related to zkSync in the codebase.

### Detailed summary
- Added zkSync Extensions
- Updated sidebar links for zkSync
- Added zkSync chain definitions
- Added zkSync custom chain data
- Updated chainConfig with getEip712Domain
- Added InvalidEip712TransactionError and related changes
- Added greeterContract and related changes
- Added isEIP712Transaction utility function
- Updated sidebar with zkSync chains
- Added Chains page for zkSync
- Added assertEip712Request and related changes
- Updated chain types with EIP712Domain
- Updated block and transaction types for zkSync
- Added assertEip712Request test
- Updated formatters for zkSync

> The following files were skipped due to too many changes: `src/chains/zksync/formatters.ts`, `src/chains/zksync/utils/isEip712Transaction.test.ts`, `src/chains/zksync/formatters.test-d.ts`, `src/clients/createClient.ts`, `test/globalSetup.ts`, `src/chains/zksync/utils/assertEip712Transaction.ts`, `src/chains/zksync/types/log.ts`, `site/pages/zksync/client.md`, `src/chains/zksync/actions/sendTransaction.test.ts`, `src/chains/zksync/index.ts`, `test/src/zksync.ts`, `src/chains/zksync/actions/signTransaction.test.ts`, `src/chains/zksync/formatters.test.ts`, `src/chains/zksync/utils/getEip712Domain.ts`, `src/chains/zksync/utils/assertEip712Transaction.test.ts`, `site/pages/zksync.mdx`, `src/chains/zksync/actions/sendEip712Transaction.test.ts`, `src/chains/zksync/actions/signEip712Transaction.test.ts`, `src/chains/zksync/serializers.ts`, `src/chains/zksync/actions/sendTransaction.ts`, `src/chains/zksync/decorators/eip712.test.ts`, `src/chains/zksync/actions/signTransaction.ts`, `src/chains/zksync/utils/getEip712Domain.test.ts`, `src/chains/zksync/actions/sendEip712Transaction.ts`, `src/chains/zksync/actions/signEip712Transaction.ts`, `src/chains/zksync/types.ts`, `site/pages/zksync/actions/sendTransaction.md`, `site/pages/zksync/actions/signTransaction.md`, `src/chains/zksync/decorators/eip712.ts`, `site/pages/zksync/actions/writeContract.md`, `src/chains/zksync/serializers.test.ts`

> ✨ Ask PR-Codex anything about this PR by commenting with `/codex {your question}`

### CODE REVIEW

File: src/chains/zksync/actions/signTransaction.ts [hunk 1]
```diff
@@ -0,0 +1,168 @@
...
+    // If we have the customSignature we can sign the transaction, doesn't matter if account type
+    // is `local` or `json-rpc`.
+    return client.chain?.serializers?.transaction(
+      {
+        chainId,
+        ...transaction,
+        customSignature,
+        type: 'eip712',
+      },
+      // Use this blank private key, probably we should change the code to be optional,
```
Comment (jxom) [inline src/chains/zksync/actions/signTransaction.ts:155]: What does this mean?

**Review by jxom — COMMENTED** [review]
Can we please have test coverage over the actions? We aim for 100% coverage.

File: src/chains/zksync/serializers.ts [hunk 2]
```diff
@@ -64,7 +64,7 @@ function serializeTransactionEIP712(
     gas ? toHex(gas) : '0x',
     to ?? '0x',
     value ? toHex(value) : '0x',
-    data ?? '0x',
+    data ?? '0x01', //data must be even number of digits
```
[the remark this answers is not in the document]
Comment (mariopil) [reply src/chains/zksync/serializers.ts]: With previous value there was an issue when a simple transaction was sent (not a contract) with empty data - server was returning a JSON RPC validation error. I've tested this once more now and realized `0x0` is enough.

File: src/chains/zksync/formatters.ts [hunk 3]
```diff
@@ -140,9 +140,11 @@ export const formatters = {
...
           },
-          type: args.type === 'eip712' ? '0x71' : '0xff',
+          type: '0x71',
         } as ZkSyncRpcTransactionRequest
       return {} as ZkSyncRpcTransactionRequest
     },
   }),
 } as const satisfies ChainFormatters
+
+// eth_call needs customSignature to use 'Array.from(hexToBytes()'
```
[the remark this answers is not in the document]
Comment (mariopil) [reply src/chains/zksync/formatters.ts]: An old comment which I forgot to remove.

File: src/chains/zksync/eip712signers.ts [hunk 4]
```diff
@@ -0,0 +1,97 @@
...
+    from: BigInt(from),
+    to: to ? BigInt(to) : 0n,
+    gasLimit: gas ?? 0n,
+    gasPerPubdataByteLimit: gasPerPubdata ?? 0n,
+    maxFeePerGas: maxFeePerGas ?? 0n,
+    maxPriorityFeePerGas: maxPriorityFeePerGas ?? 0n,
+    paymaster: paymaster ? BigInt(paymaster) : 0n,
+    nonce: nonce ? BigInt(nonce) : 0n,
+    value: value ?? 0n,
+    data: data ? data : '0x01',
```
[the remark this answers is not in the document]
Comment (mariopil) [reply src/chains/zksync/eip712signers.ts]: Changed to `0x0`

File: src/chains/zksync/actions/signTransaction.ts (hunk 1 above)
Comment (mariopil) [reply src/chains/zksync/actions/signTransaction.ts]: This comment was left by David who started this implementation back in October. I'm not sure what he exactly meant with it, I've removed it.

**Comment by mariopil** [comment]
I've added more tests, there is now 100% coverage for all zkSync related actions.

File: site/pages/zksync.mdx [hunk 5]
```diff
@@ -0,0 +1,78 @@
...
+import { eip712WalletActions } from 'viem/zksync'
+
+const walletClient = createWalletClient({ // [!code hl]
+  chain: zkSync, // [!code hl]
+  transport: custom(window.ethereum!), // [!code hl]
+}).extend(eip712WalletActions()) // [!code hl]
+```
+
+:::note[Note]
+Viem currently only supports Wallet-based Actions for zkSync.
```
Comment (dutterbutter) [inline site/pages/zksync.mdx:31]: Public actions also work as zkSync supports the standard [Ethereum JSON-RPC API](https://ethereum.org/en/developers/docs/apis/json-rpc/).

**Comment by ennioVisco** [comment]
Thanks a lot for the nice work! 
One thing that is not clear to me from the code is whether this PR takes into account zkSync Era or not.
My guess is that it doesn't, but since zkSync is now advertising it a lot, it might confuse new user that perhaps don't even know that previously zkSync worked differently.
All this just to tell that maybe it's worth adding a warning at the beginning of the zksync docs here in viem 😄

File: site/pages/zksync.mdx (hunk 5 above)
Comment (jxom) [reply site/pages/zksync.mdx:31]: This is meant to be custom zkSync Actions. Will change.

**Pull request merged**